packages/eslint-plugin/src/util/propertyTypes.ts

```diff
@@ -0,0 +1,25 @@
+import * as ts from 'typescript';
+
+import { nullThrows, NullThrowsReasons } from './nullThrows';
+
+export function getTypeOfPropertyOfType(
+  checker: ts.TypeChecker,
+  type: ts.Type,
+  propertyName: string,
+): ts.Type {
+  const reason = NullThrowsReasons.MissingToken(
+    `property "${propertyName}"`,
+    'type',
+  );
+
+  return propertyName.startsWith('__')
+    ? checker.getDeclaredTypeOfSymbol(
+        nullThrows(
+          type
+            .getProperties()
+            .find(property => property.escapedName === propertyName),
```

Comment:
Heh, this worked for the `prefer-readonly-parameter-types` rule, but is failing for `no-unnecessary-conditions` ([link](https://github.com/typescript-eslint/typescript-eslint/pull/2567/checks?check_run_id=1115991098))... will need to investigate.

Tomorrow! 😪

Reply:
Right right, the return type of the general function is intended to be `ts.Type | undefined`, and only one of the two consumers wants a `nullThrows` around it.